Fix forwarding/move bug

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53660 [experiment] AlwaysCall Backend Fallback
* #55321 Don't allocate result Tensors in out overloads: _linalg_solve_out_helper_cuda
* #53624 Boxing logic forwards arguments to stack
* **#53556 Fix forwarding/move bug**
* #53640 Don't allocate result Tensors in out overloads: at::kron_out()
* #53218 Don't allocate result Tensors in out overloads: Reduction Ops
* #53229 Fix Scalar formatting
* #53228 Fix boxing/unboxing for Scalar bool values

When packing a `Tensor&` (mutable lvalue reference) into an IValue, we accidentally didn't increase the refcount.
This wasn't triggered anywhere, until I tried to enable backend fallbacks. Backend fallbacks for ops that
have out arguments (i.e. ops that take `Tensor&` arguments and return `Tensor&` arguments) pack those returns
into an IValue stack (and accidentally don't increase the refcount), then later that stack gets destructed
(which decreases the refcount and possibly destroys the Tensor), and the `Tensor&` passed in as an out argument
is suddenty freed memory.

This PR fixes that by forwarding instead of moving when wrapping Tensors into IValues.

Differential Revision: [D26896507](https://our.internmc.facebook.com/intern/diff/D26896507/)

smessmer committed:
Fix forwarding/move bug

When packing a `Tensor&` (mutable lvalue reference) into an IValue, we accidentally didn't increase the refcount.
This wasn't triggered anywhere, until I tried to enable backend fallbacks. Backend fallbacks for ops that
have out arguments (i.e. ops that take `Tensor&` arguments and return `Tensor&` arguments) pack those returns
into an IValue stack (and accidentally don't increase the refcount), then later that stack gets destructed
(which decreases the refcount and possibly destroys the Tensor), and the `Tensor&` passed in as an out argument
is suddenty freed memory.

This PR fixes that by forwarding instead of moving when wrapping Tensors into IValues.

Differential Revision: [D26896507](https://our.internmc.facebook.com/intern/diff/D26896507/)

[ghstack-poisoned]

swolchok reviewed (approved):
Yep, `x` looks like a universal reference to me! nice catch.

smessmer committed:
Update on "Fix forwarding/move bug"


When packing a `Tensor&` (mutable lvalue reference) into an IValue, we accidentally didn't increase the refcount.
This wasn't triggered anywhere, until I tried to enable backend fallbacks. Backend fallbacks for ops that
have out arguments (i.e. ops that take `Tensor&` arguments and return `Tensor&` arguments) pack those returns
into an IValue stack (and accidentally don't increase the refcount), then later that stack gets destructed
(which decreases the refcount and possibly destroys the Tensor), and the `Tensor&` passed in as an out argument
is suddenty freed memory.

This PR fixes that by forwarding instead of moving when wrapping Tensors into IValues.

Differential Revision: [D26896507](https://our.internmc.facebook.com/intern/diff/D26896507/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix forwarding/move bug"


When packing a `Tensor&` (mutable lvalue reference) into an IValue, we accidentally didn't increase the refcount.
This wasn't triggered anywhere, until I tried to enable backend fallbacks. Backend fallbacks for ops that
have out arguments (i.e. ops that take `Tensor&` arguments and return `Tensor&` arguments) pack those returns
into an IValue stack (and accidentally don't increase the refcount), then later that stack gets destructed
(which decreases the refcount and possibly destroys the Tensor), and the `Tensor&` passed in as an out argument
is suddenty freed memory.

This PR fixes that by forwarding instead of moving when wrapping Tensors into IValues.

Differential Revision: [D26896507](https://our.internmc.facebook.com/intern/diff/D26896507/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix forwarding/move bug"


When packing a `Tensor&` (mutable lvalue reference) into an IValue, we accidentally didn't increase the refcount.
This wasn't triggered anywhere, until I tried to enable backend fallbacks. Backend fallbacks for ops that
have out arguments (i.e. ops that take `Tensor&` arguments and return `Tensor&` arguments) pack those returns
into an IValue stack (and accidentally don't increase the refcount), then later that stack gets destructed
(which decreases the refcount and possibly destroys the Tensor), and the `Tensor&` passed in as an out argument
is suddenty freed memory.

This PR fixes that by forwarding instead of moving when wrapping Tensors into IValues.

Differential Revision: [D26896507](https://our.internmc.facebook.com/intern/diff/D26896507/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix forwarding/move bug"


When packing a `Tensor&` (mutable lvalue reference) into an IValue, we accidentally didn't increase the refcount.
This wasn't triggered anywhere, until I tried to enable backend fallbacks. Backend fallbacks for ops that
have out arguments (i.e. ops that take `Tensor&` arguments and return `Tensor&` arguments) pack those returns
into an IValue stack (and accidentally don't increase the refcount), then later that stack gets destructed
(which decreases the refcount and possibly destroys the Tensor), and the `Tensor&` passed in as an out argument
is suddenty freed memory.

This PR fixes that by forwarding instead of moving when wrapping Tensors into IValues.

Differential Revision: [D26896507](https://our.internmc.facebook.com/intern/diff/D26896507/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix forwarding/move bug"


When packing a `Tensor&` (mutable lvalue reference) into an IValue, we accidentally didn't increase the refcount.
This wasn't triggered anywhere, until I tried to enable backend fallbacks. Backend fallbacks for ops that
have out arguments (i.e. ops that take `Tensor&` arguments and return `Tensor&` arguments) pack those returns
into an IValue stack (and accidentally don't increase the refcount), then later that stack gets destructed
(which decreases the refcount and possibly destroys the Tensor), and the `Tensor&` passed in as an out argument
is suddenty freed memory.

This PR fixes that by forwarding instead of moving when wrapping Tensors into IValues.

Differential Revision: [D26896507](https://our.internmc.facebook.com/intern/diff/D26896507/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix forwarding/move bug"


When packing a `Tensor&` (mutable lvalue reference) into an IValue, we accidentally didn't increase the refcount.
This wasn't triggered anywhere, until I tried to enable backend fallbacks. Backend fallbacks for ops that
have out arguments (i.e. ops that take `Tensor&` arguments and return `Tensor&` arguments) pack those returns
into an IValue stack (and accidentally don't increase the refcount), then later that stack gets destructed
(which decreases the refcount and possibly destroys the Tensor), and the `Tensor&` passed in as an out argument
is suddenty freed memory.

This PR fixes that by forwarding instead of moving when wrapping Tensors into IValues.

Differential Revision: [D26896507](https://our.internmc.facebook.com/intern/diff/D26896507/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix forwarding/move bug"


When packing a `Tensor&` (mutable lvalue reference) into an IValue, we accidentally didn't increase the refcount.
This wasn't triggered anywhere, until I tried to enable backend fallbacks. Backend fallbacks for ops that
have out arguments (i.e. ops that take `Tensor&` arguments and return `Tensor&` arguments) pack those returns
into an IValue stack (and accidentally don't increase the refcount), then later that stack gets destructed
(which decreases the refcount and possibly destroys the Tensor), and the `Tensor&` passed in as an out argument
is suddenty freed memory.

This PR fixes that by forwarding instead of moving when wrapping Tensors into IValues.

Differential Revision: [D26896507](https://our.internmc.facebook.com/intern/diff/D26896507/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix forwarding/move bug"


When packing a `Tensor&` (mutable lvalue reference) into an IValue, we accidentally didn't increase the refcount.
This wasn't triggered anywhere, until I tried to enable backend fallbacks. Backend fallbacks for ops that
have out arguments (i.e. ops that take `Tensor&` arguments and return `Tensor&` arguments) pack those returns
into an IValue stack (and accidentally don't increase the refcount), then later that stack gets destructed
(which decreases the refcount and possibly destroys the Tensor), and the `Tensor&` passed in as an out argument
is suddenty freed memory.

This PR fixes that by forwarding instead of moving when wrapping Tensors into IValues.

Differential Revision: [D26896507](https://our.internmc.facebook.com/intern/diff/D26896507/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix forwarding/move bug"


When packing a `Tensor&` (mutable lvalue reference) into an IValue, we accidentally didn't increase the refcount.
This wasn't triggered anywhere, until I tried to enable backend fallbacks. Backend fallbacks for ops that
have out arguments (i.e. ops that take `Tensor&` arguments and return `Tensor&` arguments) pack those returns
into an IValue stack (and accidentally don't increase the refcount), then later that stack gets destructed
(which decreases the refcount and possibly destroys the Tensor), and the `Tensor&` passed in as an out argument
is suddenty freed memory.

This PR fixes that by forwarding instead of moving when wrapping Tensors into IValues.

Differential Revision: [D26896507](https://our.internmc.facebook.com/intern/diff/D26896507/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix forwarding/move bug"


When packing a `Tensor&` (mutable lvalue reference) into an IValue, we accidentally didn't increase the refcount.
This wasn't triggered anywhere, until I tried to enable backend fallbacks. Backend fallbacks for ops that
have out arguments (i.e. ops that take `Tensor&` arguments and return `Tensor&` arguments) pack those returns
into an IValue stack (and accidentally don't increase the refcount), then later that stack gets destructed
(which decreases the refcount and possibly destroys the Tensor), and the `Tensor&` passed in as an out argument
is suddenty freed memory.

This PR fixes that by forwarding instead of moving when wrapping Tensors into IValues.

Differential Revision: [D26896507](https://our.internmc.facebook.com/intern/diff/D26896507/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix forwarding/move bug"


When packing a `Tensor&` (mutable lvalue reference) into an IValue, we accidentally didn't increase the refcount.
This wasn't triggered anywhere, until I tried to enable backend fallbacks. Backend fallbacks for ops that
have out arguments (i.e. ops that take `Tensor&` arguments and return `Tensor&` arguments) pack those returns
into an IValue stack (and accidentally don't increase the refcount), then later that stack gets destructed
(which decreases the refcount and possibly destroys the Tensor), and the `Tensor&` passed in as an out argument
is suddenty freed memory.

This PR fixes that by forwarding instead of moving when wrapping Tensors into IValues.

Differential Revision: [D26896507](https://our.internmc.facebook.com/intern/diff/D26896507/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix forwarding/move bug"


When packing a `Tensor&` (mutable lvalue reference) into an IValue, we accidentally didn't increase the refcount.
This wasn't triggered anywhere, until I tried to enable backend fallbacks. Backend fallbacks for ops that
have out arguments (i.e. ops that take `Tensor&` arguments and return `Tensor&` arguments) pack those returns
into an IValue stack (and accidentally don't increase the refcount), then later that stack gets destructed
(which decreases the refcount and possibly destroys the Tensor), and the `Tensor&` passed in as an out argument
is suddenty freed memory.

This PR fixes that by forwarding instead of moving when wrapping Tensors into IValues.

Differential Revision: [D26896507](https://our.internmc.facebook.com/intern/diff/D26896507/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix forwarding/move bug"


When packing a `Tensor&` (mutable lvalue reference) into an IValue, we accidentally didn't increase the refcount.
This wasn't triggered anywhere, until I tried to enable backend fallbacks. Backend fallbacks for ops that
have out arguments (i.e. ops that take `Tensor&` arguments and return `Tensor&` arguments) pack those returns
into an IValue stack (and accidentally don't increase the refcount), then later that stack gets destructed
(which decreases the refcount and possibly destroys the Tensor), and the `Tensor&` passed in as an out argument
is suddenty freed memory.

This PR fixes that by forwarding instead of moving when wrapping Tensors into IValues.

Differential Revision: [D26896507](https://our.internmc.facebook.com/intern/diff/D26896507/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix forwarding/move bug"


When packing a `Tensor&` (mutable lvalue reference) into an IValue, we accidentally didn't increase the refcount.
This wasn't triggered anywhere, until I tried to enable backend fallbacks. Backend fallbacks for ops that
have out arguments (i.e. ops that take `Tensor&` arguments and return `Tensor&` arguments) pack those returns
into an IValue stack (and accidentally don't increase the refcount), then later that stack gets destructed
(which decreases the refcount and possibly destroys the Tensor), and the `Tensor&` passed in as an out argument
is suddenty freed memory.

This PR fixes that by forwarding instead of moving when wrapping Tensors into IValues.

Differential Revision: [D26896507](https://our.internmc.facebook.com/intern/diff/D26896507/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix forwarding/move bug"


When packing a `Tensor&` (mutable lvalue reference) into an IValue, we accidentally didn't increase the refcount.
This wasn't triggered anywhere, until I tried to enable backend fallbacks. Backend fallbacks for ops that
have out arguments (i.e. ops that take `Tensor&` arguments and return `Tensor&` arguments) pack those returns
into an IValue stack (and accidentally don't increase the refcount), then later that stack gets destructed
(which decreases the refcount and possibly destroys the Tensor), and the `Tensor&` passed in as an out argument
is suddenty freed memory.

This PR fixes that by forwarding instead of moving when wrapping Tensors into IValues.

Differential Revision: [D26896507](https://our.internmc.facebook.com/intern/diff/D26896507/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix forwarding/move bug"


When packing a `Tensor&` (mutable lvalue reference) into an IValue, we accidentally didn't increase the refcount.
This wasn't triggered anywhere, until I tried to enable backend fallbacks. Backend fallbacks for ops that
have out arguments (i.e. ops that take `Tensor&` arguments and return `Tensor&` arguments) pack those returns
into an IValue stack (and accidentally don't increase the refcount), then later that stack gets destructed
(which decreases the refcount and possibly destroys the Tensor), and the `Tensor&` passed in as an out argument
is suddenty freed memory.

This PR fixes that by forwarding instead of moving when wrapping Tensors into IValues.

Differential Revision: [D26896507](https://our.internmc.facebook.com/intern/diff/D26896507/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix forwarding/move bug"


When packing a `Tensor&` (mutable lvalue reference) into an IValue, we accidentally didn't increase the refcount.
This wasn't triggered anywhere, until I tried to enable backend fallbacks. Backend fallbacks for ops that
have out arguments (i.e. ops that take `Tensor&` arguments and return `Tensor&` arguments) pack those returns
into an IValue stack (and accidentally don't increase the refcount), then later that stack gets destructed
(which decreases the refcount and possibly destroys the Tensor), and the `Tensor&` passed in as an out argument
is suddenty freed memory.

This PR fixes that by forwarding instead of moving when wrapping Tensors into IValues.

Differential Revision: [D26896507](https://our.internmc.facebook.com/intern/diff/D26896507/)

[ghstack-poisoned]

smessmer committed:
Update on "Fix forwarding/move bug"


When packing a `Tensor&` (mutable lvalue reference) into an IValue, we accidentally didn't increase the refcount.
This wasn't triggered anywhere, until I tried to enable backend fallbacks. Backend fallbacks for ops that
have out arguments (i.e. ops that take `Tensor&` arguments and return `Tensor&` arguments) pack those returns
into an IValue stack (and accidentally don't increase the refcount), then later that stack gets destructed
(which decreases the refcount and possibly destroys the Tensor), and the `Tensor&` passed in as an out argument
is suddenty freed memory.

This PR fixes that by forwarding instead of moving when wrapping Tensors into IValues.

Differential Revision: [D26896507](https://our.internmc.facebook.com/intern/diff/D26896507/)

[ghstack-poisoned]
